content/blog/2025-12-16-allo-xls-backend.md

```diff
@@ -0,0 +1,218 @@
+
+Where DSLX preserves loop structure through functional for-expressions, XLS IR eliminates loops entirely through unrolling. Multi-dimensional arrays are represented as one-dimensional arrays in XLS IR, requiring explicit linearization. For a 2D memref of shape N×M, a load `affine.load %A[%i, %j]` translates to computing `linear_idx = i * M + j`, then `array_index(A, indices=[linear_idx])`. Array updates are more intricate, and for each store we generate a fresh array name, and the lowerer must rebind all subsequent references.
+
+### Systolic Array Generation (Method 1)
+
+XLS organizes computation as communicating processes based on Kahn process network semantics. A `proc` represents a stateful concurrent process with three components:
+- A `config` function declaring and wiring communication channels
+- An `init` function establishing initial state values  
+- A `next` function implementing a single state transition by consuming inputs, updating state, and producing outputs
+
+Rather than attempt general lowering from arbitrary Allo programs to explicit state machines (an intractably hard problem), we adopt a pattern-matching approach that specifically recognizes systolic array structures in MLIR and generates corresponding XLS grid implementations. We extracted metadata from the MLIR that drives a builder. The builder constructs DSLX proc ASTs for the PE and for the systolic grid that is then serialized.
+
+To see the generated XLS, see this link: https://github.com/Nikil-Shyamsunder/allo-xls-backend/tree/main/allo/backend/xls/examples/systolic
+
+### Meta-Systolic Arrays (Method 2)
+
+Beyond the basic pattern-matching approach for library-style systolic arrays, we developed a lowering strategy for systolic arrays expressed using Allo's metaprogramming constructs. However, the key challenge is that Allo's `meta_if` constructs are compiled away before we see them—by the time our compiler receives the MLIR, the conditional code generation has already happened.
+
+When Allo compiles a dataflow kernel like:
+```python
+@df.kernel(mapping=[P0, P1])
+def gemm(A, B, C):
+    i, j = df.get_pid()
+    with allo.meta_if(i == 0):
+        # Load B columns
+        ...
+    with allo.meta_elif(j == 0):
+        # Load A rows
+        ...
+    with allo.meta_else():
+        # Interior PE: MAC computation
+        ...
+```
+
+The resulting MLIR contains separate functions for each grid position: `gemm_0_0`, `gemm_0_1`, `gemm_1_0`, `gemm_1_1`, etc. Each function has different code depending on which `meta_if` branch was selected for that `(i, j)` coordinate at compile time. The challenge of this approach was that we had to reverse-engineer the spatial structure and PE types from this unrolled representation by analyzing the behavior of each generated function.
```

Comment:
Hmm, I don't quite understand why you need to "reverse-engineer the spatial structure and PE types." First of all, I'm not exactly sure what "spatial structure" means here… but assuming it's something about the 2D layout of the PEs, I'm also not sure why that matters. I guess I'd assume that the PEs already communicate through some manner of communication channel, and that's all you need to "link them up" according to their data exchange?